group/edwards25519/point.go

```diff
@@ -290,3 +296,319 @@ func (P *point) IsCanonical(s []byte) bool {
 
 	return 1-(c&d&1) == 1
 }
+
+func (P *point) Hash(m []byte, dst string) kyber.Point {
+	u := hashToField(m, dst, 2)
+	q0 := mapToCurveElligator2Ed25519(u[0])
+	q1 := mapToCurveElligator2Ed25519(u[1])
+	P.Add(q0, q1)
+
+	// Clear cofactor
+	P.Mul(cofactorScalar, P)
+
+	return P
+}
+
+func hashToField(m []byte, dst string, count int) []fieldElement {
+	l := 48 // L param in RFC9380
+	byteLen := count * l
+	uniformBytes, _ := expandMessageXMD(sha512.New(), m, dst, byteLen)
+
+	u := make([]fieldElement, count)
+	for i := 0; i < count; i++ {
+		elmOffset := l * i
+		tv := big.NewInt(0).SetBytes(uniformBytes[elmOffset : elmOffset+l])
+		tv = tv.Mod(tv, prime)
+		fe := fieldElement{}
+		feFromBn(&fe, tv)
+		u[i] = fe
+	}
+
+	return u
+}
+
+func expandMessageXMD(h hash.Hash, m []byte, domainSeparator string, byteLen int) ([]byte, error) {
+	r := float64(byteLen) / float64(h.Size()>>3)
+	ell := int(math.Ceil(r))
+	if ell > 255 || ell < 0 || byteLen > 65535 {
+		return nil, errors.New("invalid parameters")
+	}
+
+	if len(domainSeparator) > 255 {
+		longDstSep := "H2C-OVERSIZE-DST-"
+		h.Reset()
+		h.Write([]byte(longDstSep))
+		h.Write([]byte(domainSeparator))
+
+		domainSeparator = string(h.Sum(nil))
+	}
+
+	padDom, err := i2OSP(len(domainSeparator), 1)
+	if err != nil {
+		return nil, err
+	}
+
+	dstPrime := append([]byte(domainSeparator), padDom...)
+	byteLenStr, _ := i2OSP(byteLen, 2)
+	zeroPad, _ := i2OSP(0, 1)
+	zPad, _ := i2OSP(0, h.BlockSize())
+
+	// mPrime = Z_pad || msg || l_i_b_str || I2OSP(0, 1) || DST_prim
+	mPrime := make([]byte, 0, len(zPad)+len(m)+len(byteLenStr)+len(zeroPad)+len(dstPrime))
+	mPrime = append(mPrime, zPad...)
+	mPrime = append(mPrime, m...)
+	mPrime = append(mPrime, byteLenStr...)
+	mPrime = append(mPrime, zeroPad...)
+	mPrime = append(mPrime, dstPrime...)
+
+	// b0 = H(msg_prime)
+	h.Reset()
+	h.Write([]byte(mPrime))
+	b0 := h.Sum(nil)
+
+	// b_1 = H(b_0 || I2OSP(1, 1) || DST_prime
+	h.Reset()
+	h.Write(b0)
+	onePad, _ := i2OSP(1, 1)
+	h.Write([]byte(onePad))
+	h.Write([]byte(dstPrime))
+	b1 := h.Sum(nil)
+
+	bFinal := make([]byte, 0, len(b1)*(ell+1))
+	bFinal = append(bFinal, b1...)
+	bPred := b1
+	for i := 2; i <= ell; i++ {
+		x, err := byteXor(bPred, b0, bPred)
+		if err != nil {
+			return nil, err
+		}
+		ithPad, _ := i2OSP(i, 1)
+
+		h.Reset()
+		h.Write(x)
+		h.Write(ithPad)
+		h.Write(dstPrime)
+
+		bPred = h.Sum(nil)
+		bFinal = append(bFinal, bPred...)
+	}
+
+	return bFinal[:byteLen], nil
+}
+
+func expandMessageXOF(h sha3.ShakeHash, m []byte, domainSeparator string, byteLen int) ([]byte, error) {
+	if byteLen > 65535 {
+		return nil, errors.New("invalid parameters")
+	}
+
+	if len(domainSeparator) > 255 {
```

Comment:
You have this twice, maybe create a function that handles long domain separators?

Reply:
The second one takes a different type of hash function, I don't think it make much sense to have a function for that, I did de-duplicated the repeated string though.